be/src/storage/tablet_updates.cpp

```diff
@@ -2620,6 +2624,173 @@ Status TabletUpdates::compaction(MemTracker* mem_tracker) {
+                                                 _tablet.tablet_id(), rowsetid);
+                DCHECK(false) << msg;
+                LOG(WARNING) << msg;
+            } else if (itr->second->compaction_score > 0) {
+                auto& stat = *itr->second;
+                total_valid_rowsets++;
+                total_valid_segments += stat.num_segments;
+                int32_t level = _calc_compaction_level(&stat);
+                candidates_by_level[level].emplace_back();
+                auto& e = candidates_by_level[level].back();
+                e.rowsetid = itr->first;
+                e.score_per_row =
+                        (level == -1) ? 0 : (float)((double)stat.compaction_score / (stat.num_rows - stat.num_dels));
+                e.num_rows = stat.num_rows;
+                e.num_dels = stat.num_dels;
+                e.bytes = stat.byte_size;
+                e.num_segments = stat.num_segments;
+            }
+        }
+    }
+
+    int64_t total_rows = 0;
+    int64_t total_bytes = 0;
+    int32_t compaction_level = -1;
+    int64_t max_score = 0;
+    for (auto& [level, candidates] : candidates_by_level) {
+        if (level == -1) {
+            continue;
+        }
+        int64_t total_segments = 0;
+        int64_t del_rows = 0;
+        int64_t level_score = 0;
+        for (auto& e : candidates) {
+            level_score += e.score_per_row * (e.num_rows - e.num_dels);
+            total_segments += e.num_segments;
+            del_rows += e.num_dels;
+        }
+        if (level_score > max_score && (total_segments > 1 || del_rows > 0)) {
+            compaction_level = level;
+            max_score = level_score;
+        }
+    }
+
+    int64_t total_merged_segments = 0;
+    RowsetStats stat;
+    std::set<int32_t> compaction_level_candidate;
+    do {
+        auto iter = candidates_by_level.find(compaction_level);
+        if (iter == candidates_by_level.end()) {
+            break;
+        }
+        for (auto& e : iter->second) {
+            size_t new_rows = stat.num_rows + e.num_rows - e.num_dels;
+            size_t new_bytes = stat.byte_size;
+            if (e.num_rows != 0) {
+                new_bytes += e.bytes * (e.num_rows - e.num_dels) / e.num_rows;
+            }
+            if ((stat.byte_size > 0 && new_bytes > config::update_compaction_result_bytes * 2) ||
+                info->inputs.size() >= config::max_update_compaction_num_singleton_deltas) {
+                break;
+            }
+            max_score += e.score_per_row * (e.num_rows - e.num_dels);
+            info->inputs.emplace_back(e.rowsetid);
+            stat.num_rows = new_rows;
+            stat.byte_size = new_bytes;
+            total_rows += e.num_rows;
+            total_bytes += e.bytes;
+            total_merged_segments += e.num_segments;
+        }
+        compaction_level_candidate.insert(compaction_level);
+        compaction_level = _calc_compaction_level(&stat);
```

Comment:
only merge level when the compaction_level is the next max score level?

Reply:
> only merge level when the compaction_level is the next max score level?

No. If the new rowset is level `x`, the level `x` rowsets will be picked to this compaction.